chore(engine): notify when jobs are available

deepthidevaki opened the pull request:
# Description
`ActivatableJobsCount` maintains an in-memory map of jobtype to activatable job count. When stream processor is opened and recovered, the map is initialized by iterating over the RocksDB state. After that `JobState` notifies when a job is created or marked not activatable. When the count is updated from 0 to 1, `onJobsAvailableCallback` is invoked which broadcasts `jobType`.

# Related issues
Related to #2828

closes #2824

# Pull Request Checklist

- [x] I have signed the Camunda CLA
- [x] I followed the [contributor guidelines](/CONTRIBUTING.md)
- [x] All commit messages match our [commit message guidelines](/CONTRIBUTING.md#commit-message-guidelines)
- [x] If submitting code, please run `mvn clean install -DskipTests` locally

saig0 commented on engine/src/test/java/io/zeebe/engine/processor/workflow/job/ActivatableJobsNotificationTests.java:
```diff
@@ -0,0 +1,149 @@
+                .limit(amount)
+                .count())
+        .isEqualTo(amount);
+    return instanceKeys;
```
```suggestion
    return IntStream.range(0, amount)
      .mapToObj(i -> ENGINE.createJob(taskType, PROCESS_ID))
      .map(r -> r.getValue().getWorkflowInstanceKey())
      .collect(Collectors.toList());
```

We can use our existing logic here.

saig0 commented on engine/src/test/java/io/zeebe/engine/processor/workflow/job/ActivatableJobsNotificationTests.java:
```diff
@@ -0,0 +1,149 @@
+import org.junit.Test;
+import org.mockito.Mockito;
+
+public class ActivatableJobsNotificationTests {
```
We could improve the coverage by adding test cases for:
* job timed out
* job failed with retries > 0
* job get resolved after failed with retries = 0
* multiple job types

deepthidevaki commented:
@saig0 I have updated the PR.
* Removed the count, instead check if it is the first job activated by reading the rocksdb state and notify.
* Added `existsPrefix()` to the `ColumnFamily`
* Added more tests.

saig0 commented on zb-db/src/main/java/io/zeebe/db/impl/rocksdb/transaction/TransactionalColumnFamily.java:
```diff
@@ -148,6 +149,18 @@ public boolean exists(DbContext context, KeyType key) {
+          setPrefixExists(true);
+          return false;
+        });
+    return prefixExists;
```
Maybe, we should move this logic into `ZeebeTransactionDb` nex to the other database accesses.

deepthidevaki replied:
Agree :+1: Moved and added some tests.

deepthidevaki commented:
bors r+
